File: leetcode/39.combination-sum.py
# @leet start
class Solution:
    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        candidates.sort()

        res: list[list[int]] = []
        curr: list[int] = []

        def backtrack(i: int, accum: int) -> None:
            """i is candidate index, accum is sum of curr."""
            if accum == target:
                res.append(curr[:])
                return

            for j in range(i, len(candidates)):
                new_accum = accum + candidates[j]

                # Early return optimization
                if new_accum > target:
                    break

                curr.append(candidates[j])
                backtrack(j, new_accum)
                curr.pop()

        backtrack(0, 0)

        return res


# @leet end

File: leetcode/39.combination-sum.ts
// @leet start
function combinationSum(candidates: number[], target: number): number[][] {
  const result: number[][] = [];
  const curr: number[] = [];

  const backtrack = (i: number, amount: number) => {
    if (amount >= target) {
      if (amount === target) {
        result.push([...curr]);
      }

      return;
    }

    for (let j = i; j < candidates.length; j++) {
      curr.push(candidates[j]);
      backtrack(j, amount + candidates[j]);
      curr.pop();
    }
  };

  backtrack(0, 0);

  return result;
}
// @leet end
